extensions-contrib/moving-average-query/src/main/java/org/apache/druid/query/movingaverage/MovingAverageIterable.java

```diff
@@ -0,0 +1,301 @@
+ * See {@link MovingAverageIterator#computeMovingAverage(Map, Row, boolean)} for more details.
+ */
+public class MovingAverageIterable implements Iterable<Row>
+{
+
+  private final Sequence<RowBucket> seq;
+  private final Collection<DimensionSpec> dims;
+  private final Collection<AveragerFactory<?, ?>> factories;
+  private final Map<String, PostAggregator> postAggMap;
+  private final Map<String, AggregatorFactory> aggMap;
+  private final Map<String, Object> fakeEvents;
+
+  public MovingAverageIterable(
+      Sequence<RowBucket> buckets,
+      Collection<DimensionSpec> dims,
+      Collection<AveragerFactory<?, ?>> factories,
+      List<PostAggregator> postAggList,
+      List<AggregatorFactory> aggList
+  )
+  {
+    this.dims = dims;
+    this.factories = factories;
+    this.seq = buckets;
+
+    postAggMap = postAggList.stream().collect(Collectors.toMap(postAgg -> postAgg.getName(), postAgg -> postAgg));
+    aggMap = aggList.stream().collect(Collectors.toMap(agg -> agg.getName(), agg -> agg));
+
+    ColumnSelectorFactory colFact = new ColumnSelectorFactory()
+    {
+      @Override
+      public DimensionSelector makeDimensionSelector(DimensionSpec dimensionSpec)
+      {
+        // Generating empty records while aggregating on Filtered aggregators requires a dimension selector
+        // for initialization.  This dimension selector is not actually used for generating values
+        return DimensionSelector.constant(null);
+      }
+
+      @Override
+      public ColumnValueSelector makeColumnValueSelector(String s)
+      {
+        return null;
+      }
+
+      @Override
+      public ColumnCapabilities getColumnCapabilities(String s)
+      {
+        return null;
+      }
+    };
+    // Fill in all the fake events
+    fakeEvents = new LinkedHashMap<>();
+    aggMap.values().forEach(agg -> {
+      Aggregator aggFactorized = agg.factorize(colFact);
+      fakeEvents.put(agg.getName(), aggFactorized.get());
+    });
+    postAggMap.values().forEach(postAgg -> fakeEvents.put(postAgg.getName(), postAgg.compute(fakeEvents)));
+  }
+
+  /* (non-Javadoc)
+   * @see java.lang.Iterable#iterator()
+   */
+  @Override
+  public Iterator<Row> iterator()
+  {
+    return new MovingAverageIterator(seq, dims, factories, fakeEvents, aggMap);
+  }
+
+  static class MovingAverageIterator implements Iterator<Row>
+  {
+
+    private final Collection<DimensionSpec> dims;
+    private final Map<Map<String, Object>, Collection<Averager<?>>> averagers = new HashMap<>();
+    private final Collection<AveragerFactory<?, ?>> factories;
+
+    private Yielder<RowBucket> yielder;
+    private RowBucket cache = null;
+    private Iterator<Row> cacheIter;
+    private Iterator<Map<String, Object>> averagersKeysIter;
+    private Set<Map<String, Object>> seenKeys = new HashSet<>();
+    private Row saveNext;
+    private Map<String, AggregatorFactory> aggMap;
+    private Map<String, Object> fakeEvents;
```

Comment:
What is `fakeEvents`?

Reply:
Added a comment (in new pr, see commit `2419578`).